de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUPackingMaterialDAO.java

```diff
@@ -51,10 +59,48 @@
 	@Override
 	public I_M_HU_PackingMaterial retrivePackingMaterialOfProduct(final I_M_Product product)
 	{
-		return Services.get(IQueryBL.class).createQueryBuilder(I_M_HU_PackingMaterial.class, product)
+		return iQueryBL.createQueryBuilder(I_M_HU_PackingMaterial.class, product)
 				.filter(new EqualsQueryFilter<I_M_HU_PackingMaterial>(I_M_HU_PackingMaterial.COLUMNNAME_M_Product_ID, product.getM_Product_ID()))
 				.create()
 				.setOnlyActiveRecords(true)
 				.firstOnly(I_M_HU_PackingMaterial.class);
 	}
+
+	@Nullable
+	@Override
+	public I_M_HU_PackingMaterial retrievePackingMaterialOrNull(@NonNull final PackageId packageId)
+	{
+		return iQueryBL
+				.createQueryBuilder(I_M_Package_HU.class)
+				.addEqualsFilter(I_M_Package_HU.COLUMNNAME_M_Package_ID, packageId)
+				//
+
+				.andCollect(I_M_HU.COLUMN_M_HU_ID, I_M_HU.class)
+				.andCollectChildren(I_M_HU_Item.COLUMN_M_HU_ID)
+				.andCollect(I_M_HU_PackingMaterial.COLUMN_M_HU_PackingMaterial_ID, I_M_HU_PackingMaterial.class)
+				.create()
+				.first(I_M_HU_PackingMaterial.class);
+	}
+
+	@Nullable
+	@Override
+	public I_M_HU_PackingMaterial getHUPackingMaterial(@NonNull final I_M_HU_Item huItem)
+	{
+		final int packingMaterialId = huItem.getM_HU_PackingMaterial_ID();
+		return packingMaterialId > 0 ? loadOutOfTrx(packingMaterialId, I_M_HU_PackingMaterial.class) : null;
+	}
+
+	@NonNull
+	@Override
+	public PackageDimensions preparePackageDimensions(@NonNull final I_M_HU_PackingMaterial packingMaterial, @NonNull final UomId toUomId)
+	{
+		final UomId fromUomId = UomId.ofRepoId(packingMaterial.getC_UOM_Dimension_ID());
+
+		final IUOMConversionBL iuomConversionBL = Services.get(IUOMConversionBL.class);
+		return PackageDimensions.builder()
+				.heightInCM(iuomConversionBL.convert(fromUomId, toUomId, packingMaterial.getHeight()).orElse(BigDecimal.valueOf(1)).intValue())
```

Comment:
I believe this will make my implementation almost as ugly.
Compare

```java
.heightInCM(iuomConversionBL.convertQtyTo(Quantitys.create(packingMaterial.getHeight(), fromUomId), toUomId).map(qty -> qty.toBigDecimal()).orElse(BigDecimal.ONE).intValue())
```

with

```java
.heightInCM(iuomConversionBL.convert(fromUomId, toUomId, packingMaterial.getHeight()).orElse(BigDecimal.ONE).intValue())
```

I still won't use `Quantity` calls here.

Btw, thanks for pointing out `Optional.map` to me!, never used it before :)